x/mongo/driver/topology/pool.go

```diff
@@ -710,11 +884,23 @@ func (p *pool) maintain(ctx context.Context, wg *sync.WaitGroup) {
 	wantConns := make([]*wantConn, 0, p.minSize)
 	defer func() {
 		for _, w := range wantConns {
-			w.tryDeliver(nil, ErrPoolDisconnected)
+			w.tryDeliver(nil, ErrPoolClosed)
 		}
 	}()
 
 	for {
+		select {
+		case <-ticker.C:
+		case <-p.maintainReady:
+		case <-ctx.Done():
+			return
+		}
+
+		// Only maintain the pool while it's in the "ready" state.
+		if p.getState() != poolReady {
+			continue
+		}
+
```

Comment:
This came up during our meeting. If the pool is cleared just after L900, `maintain` may try to rebuild the pool. It seems unlikely to occur. The `maintainInterval` is 10 seconds and not configurable by public API.

I think the worst case is when `minPoolSize` is large and `maxConnecting` is large. For example, suppose `minPoolSize` is 1000 and `maxConnecting` is 100:
- 1000 `wantConn` are added to `newConnWait`.
- Each of the 100 `createConnections` creates a connection.
   - If any fail to connect, the pool will be cleared.
   - If all succeed at connecting. They will be checked in and unused until the server state is updated to something aside from "Unknown".

The impact does not seem too harmful to me. Consider adding a comment to describe this possible scenario.

Reply:
Actually I realized I solved the same race condition in `checkOut()` by holding the `stateMu` read lock when entering the wait queue, which prevents state changes between checking the state and entering the wait queue. I've updated `maintain()` to hold the `stateMu` read lock while modifying the wait queue as well.

Reply:
Oh nice, that is good to it is capped at 10 new connections.